src/cosalib/cmdlib.py

```diff
@@ -75,10 +75,34 @@ def get_lock_path(path):
     return os.path.join(dn, f".{bn}.lock")
 
 
-def write_json(path, data, lock_path=None):
+# Credit to @artix
+def merge_dicts(x, y):
+    """
+    Merge two dicts recursively, but based on the difference.
+    """
+    sd = set(x.keys()).symmetric_difference(y.keys())
+    ret = {}
+    for d in [x, y]:
+        for k, v in d.items():
+            if k in sd:
+                # the key is only present in one dict, add it directly
+                ret.update({k: v})
+            elif type(x[k]) == dict and type(y[k]) == dict:
+                # recursively merge
+                ret.update({k: merge_dicts(x[k], y[k])})
+            else:
+                # first dictionary always takes precedence
```

Comment:
Can we instead error out here? When would this ever happen (that an "secondary" `meta.json` wants to change a field)? I think this probably highlights some issue we need to resolve; we might be dropping new information we don't intend to.

IIUC, I think this would also then make the operation commutative, which should make the logic in `merge_meta` simpler.

Reply:
This seems like a similar thread to https://github.com/coreos/coreos-assembler/pull/1675#discussion_r474865910

Reply:
This deals with the parallel build case where two `buildextends` are operating on the same meta.json and determines which takes precedence.  For example, if `aws` and `gcp` are building at the same time, the on-disk stamp would be incremented when the first writes to meta.json; we want to know if the in-memory version loaded is newer or older than the on-disk version.

Reply:
> Can we instead error out here? When would this ever happen (that an "secondary" meta.json wants to change a field)

If the value of `k` is in both `x` and `y`.  This is demonstrated in https://github.com/coreos/coreos-assembler/pull/1675/files#diff-3ab0f5673838863cf798659fabf711b8R28. 

TBH, the `else:` could be dropped entirely.

Reply:
> I think the idea is that there should never be overlaps so this should just die out complaining that the inputs are invalid.

Right. To be more clear and tweak my original ask, this branch is taken when both `x` and `y` have some key `k` which *isn't* a dictionary. Either they have the same value, or they don't. If they don't, then that indicates that somehow a `buildextend` command wants to change a primitive field, right? Could we error out there? (One exception is probably timestamp modification, though that one is already special since we want to bump it anyway).

This is basically a more generic check than the spot checking for equality of e.g. `ostree-commit` and `ostree-content-checksum` below. (So then we'd be able to just drop that check too).

Reply:
https://github.com/darkmuggle/coreos-assembler/commit/2c67214bd72d5c4de8a4005d04f014b5eaadd352#diff-ee8072ded537bd9238ceb54ece4ea85bR111 ensures that we have the source and the destination being merged via `cmd-meta --finalize` pass the schema, and the schema check on write ensures we don't mess things up. 

Do you mind if we explore that in a follow-up PR? I'm not trying to duck the question, but I want to treat it properly. Right now the checks in place make sure we have it valid to the schema, but I have an idea of how to deal with the check.

Reply:
https://github.com/darkmuggle/coreos-assembler/commit/c80ae52bd6b7abb8efc35666d55a6761b76cc111 is a potential follow up. This would allow you do `if x == y` to see the builds are the same.

Reply:
> Do you mind if we explore that in a follow-up PR? I'm not trying to duck the question, but I want to treat it properly.

Sure, SGTM!